frontend_tests/casper_tests/04-compose.js

```diff
@@ -141,6 +141,45 @@ casper.then(function () {
     });
 });
 
+casper.waitUntilVisible('#markdown_preview', function () {
+    casper.test.assertNotVisible('#undo_markdown_preview', 'Write button is hidden');
+    casper.then(function () {
+        casper.click("#markdown_preview");
+    });
+    casper.then(function () {
+        casper.waitWhileVisible("#markdown_preview", function () {
+            casper.test.assertVisible('#undo_markdown_preview', 'Write button is visible');
+            casper.test.assertEquals(casper.getHTML('#preview_message_area'), "Nothing to preview", "Nothing to preview");
+        });
+    });
+});
+
+casper.waitUntilVisible('#undo_markdown_preview', function () {
+    casper.then(function () {
+        casper.click("#undo_markdown_preview");
+    });
+
+    casper.then(function () {
+        casper.waitWhileVisible("#undo_markdown_preview", function () {
+            casper.test.assertVisible('#markdown_preview', 'Preview button is visible.');
+            casper.test.assertNotVisible('#undo_markdown_preview', 'Write button is hidden.');
+            casper.test.assertEquals(casper.getHTML('#preview_message_area'), "", "Markdown preview area is empty");
+
+            casper.fill('form[action^="/json/messages"]', {
+                content: '**Markdown Preview** >> Test for markdown preview'
+            }, false);
+
+            casper.then(function () {
+                casper.click("#markdown_preview");
+            });
+            casper.then(function () {
+                casper.waitWhileVisible("#markdown_preview", function () {
+                    casper.test.assertEquals(casper.getHTML('#preview_message_area'), "<p><strong>Markdown Preview</strong> &gt;&gt; Test for markdown preview</p>", "Check markdown is previewed properly");
+                });
+            });
+        });
+    });
+});
```

Comment:
I am not too much sure about nesting, the documentation only says that to avoid race conditions `waitFor*` functions shouldn't be chained together. `casper.waitUntilVisible('#undo_markdown_preview')` should be wrapped in a `then` otherwise it will have a race condition with `casper.waitUntilVisible('#markdown_preview'`. Of course, race condition will only be a problem if both tests are contending for the same shared resources.